staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store_test.go

```diff
@@ -531,6 +641,343 @@ func TestNoOpUpdates(t *testing.T) {
 	}
 }
 
+func TestStoreUpdateHooks(t *testing.T) {
+	// To track which hooks were called in what order.  Not all hooks can
+	// mutate the object.
+	var milestones []string
+
+	setAnn := func(obj runtime.Object, key string) {
+		pod := obj.(*example.Pod)
+		if pod.Annotations == nil {
+			pod.Annotations = make(map[string]string)
+		}
+		pod.Annotations[key] = "true"
+	}
+	mile := func(s string) {
+		milestones = append(milestones, s)
+	}
+
+	testCases := []struct {
+		name             string
+		decorator        ObjectFunc
+		begin            func(context.Context, runtime.Object, runtime.Object, *metav1.UpdateOptions) (func(context.Context, bool), error)
+		after            ObjectFunc
+		expectError      bool
+		expectAnnotation string   // to test object mutations
+		expectMilestones []string // to test sequence
+	}{{
+		name: "no hooks",
+	}, {
+		name: "Decorator mutation",
+		decorator: func(obj runtime.Object) {
+			setAnn(obj, "DecoratorWasCalled")
+		},
+		expectAnnotation: "DecoratorWasCalled",
+	}, {
+		name: "AfterUpdate mutation",
+		after: func(obj runtime.Object) {
+			setAnn(obj, "AfterUpdateWasCalled")
+		},
+		expectAnnotation: "AfterUpdateWasCalled",
+	}, {
+		name: "BeginUpdate mutation",
+		begin: func(_ context.Context, obj, _ runtime.Object, _ *metav1.UpdateOptions) (func(context.Context, bool), error) {
+			setAnn(obj, "BeginUpdateWasCalled")
+			return func(context.Context, bool) {}, nil
+		},
+		expectAnnotation: "BeginUpdateWasCalled",
+	}, {
+		name: "success ordering",
+		decorator: func(obj runtime.Object) {
+			mile("Decorator")
+		},
+		after: func(obj runtime.Object) {
+			mile("AfterUpdate")
+		},
+		begin: func(_ context.Context, obj, _ runtime.Object, _ *metav1.UpdateOptions) (func(context.Context, bool), error) {
+			mile("BeginUpdate")
+			return func(context.Context, bool) {
+				mile("Cleanup")
```

Comment:
I'd like these tests to expand to indicate whether they were called with success true or false.  I like the test.

Reply:
ACK.  Also will cover failure paths better